File: dp/day4_jumpgame/45.jump-game-ii.py
#
# @lc app=leetcode id=45 lang=python3
#
# [45] Jump Game II
#

# @lc code=start
from functools import cache


class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        
        @cache
        def dp(i: int)-> int:
            if i>=n-1: return 0
            if i+nums[i] >=n-1: return 1
            
            ans =n
            for s in range(1, nums[i]+1):
                ans = min(ans, dp(i+s)+1)
            return ans
        return dp(0)
        
# @lc code=end

File: dp/day4_jumpgame/45.jump-game-ii.cpp
/*
 * @lc app=leetcode id=45 lang=cpp
 *
 * [45] Jump Game II
 */

#include "bits/stdc++.h"

using namespace std;


// https://leetcode.com/problems/jump-game-ii/discuss/1192401/Easy-Solutions-w-Explanation-or-Optimizations-from-Brute-Force-to-DP-to-Greedy-BFS

// @lc code=start

// button up
class Solution {
public:
  int jump(vector<int> &nums) {
    int jumps = 0, currJumpEnd = 0, farthest = 0;
    for (int i = 0; i < nums.size() - 1; ++i) {
     
    }
  }
};
// @lc code=end


// greedy BFS
class Solution {
public:
  int jump(vector<int> &nums) {
    int jumps = 0, currJumpEnd = 0, farthest = 0;
    for (int i = 0; i < nums.size() - 1; ++i) {
      farthest = max(farthest, i + nums[i]);

      if (i == currJumpEnd) {
        ++jumps;
        currJumpEnd = farthest;
      }
    }
    return jumps;
  }
};

int main() {
  Solution s;
  vector<int> ivec{2, 3, 1, 1, 4};
  cout << s.jump(ivec) << endl;
}
